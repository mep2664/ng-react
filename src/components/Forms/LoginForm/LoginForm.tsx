import * as React from "react";
import { TextInput } from "../..";
import styled from "styled-components";
import { emphasisType } from "../../../theme";

const FormHeader = styled.h2`
    text-align: center;
`;

interface ILoginFormProps {
    formId: string;
    emphasis: emphasisType;
    onSubmitChange?: (isSubmitting: boolean) => void;
}

export const LoginForm: React.FC<ILoginFormProps> = ({ formId, emphasis = "Primary", onSubmitChange }) => {
    const [email, setEmail] = React.useState<string>("");
    const [password, setPassword] = React.useState<string>("");
    const [error, setError] = React.useState<string>("");
    const [isSubmitting, setSubmitting] = React.useState<boolean>(false);

    React.useEffect(() => {
        if (onSubmitChange) {
            onSubmitChange(isSubmitting);
        }
    }, [isSubmitting, onSubmitChange])

    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        setSubmitting(true);
        const data = {
            email,
            password,
        }
        fetch("http://localhost:5556/rest/login", {
            method: "POST",
            mode: "cors",
            cache: "no-cache",
            credentials: "same-origin",
            headers: {
                "Content-Type": "application/json; charset=utf-8",
                "Accept": "application/json",
            },
            redirect: "follow",
            referrer: "no-referrer",
            body: JSON.stringify(data),
        }).then((response) => {
            if (response.ok) {
                response.json().then((session) => {
                    console.log(session);
                    const d = new Date();
                    const numHours = 4;
                    const expires = d.setTime(d.getTime() + ((numHours * 60 * 60 * 1000)));
                    document.cookie = `uuid=${session.sessionID};expires=${expires};path=/`;
                    if (window.location.pathname === "/login" || window.location.pathname === "/home" || window.location.pathname === "/") {
                        console.log("here");
                        console.log(window.location.host + "/dashboard");
                        window.location.assign(`${window.location.protocol}//${window.location.host}/dashboard`);
                    } else {
                        window.location.reload()
                    }

                }, (reason) => { setError(reason.toString()); setSubmitting(false); });
            } else {
                response.json().then((reason) => {
                    setError(reason.error);
                    setSubmitting(false);
                }, (reason) => { setError(reason.toString()); setSubmitting(false); });
            }
        }, (reason) => { setError(reason.toString()); setSubmitting(false); });
    }

    return (
        <form id={formId} onSubmit={handleSubmit}>
            <FormHeader>Login</FormHeader>
            {error && <div style={{ color: "red" }}>{error}</div>}
            <TextInput emphasis={emphasis} name="email" label="Email" value={email} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setEmail(e.target.value)} />
            <TextInput emphasis={emphasis} name="password" label="Password" value={password} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setPassword(e.target.value)} />
        </form >
    );
}