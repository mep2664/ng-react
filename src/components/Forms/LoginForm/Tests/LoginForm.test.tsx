import React from "react"
import { render, fireEvent, wait } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect";
import { LoginForm, LOGIN_USER } from "../LoginForm"
import { MockedProvider } from '@apollo/react-testing';

const vars = { email: "EMAIL@EMAILTEST.DNE", password: "PASSWORD" };
let numTimesSubmitted = 0;
const mocks = [
    {
        request: {
            query: LOGIN_USER,
            variables: { ...vars },
        },

        result: () => {
            numTimesSubmitted++;
            return {
                data: {
                    loginUser: {
                        token: "TEST_AUTHTOKEN",
                        error: "",
                    },
                },
            };
        },
    },
];

const props = {
    formId: "FORM_ID",
    onSubmitChange: jest.fn(),
};

describe("Tests for the LoginForm component", () => {
    beforeEach(() => {
        numTimesSubmitted = 0;
        jest.resetAllMocks();
    })

    test("Component matches snapshot", async () => {
        const { container } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <LoginForm {...props} />
            </MockedProvider>,
        );
        await wait();
        expect(container).toMatchSnapshot();
    });

    test("Changes email correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <LoginForm {...props} />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Email");
        fireEvent.change(input, { target: { value: vars.email } });
        expect(input).toHaveValue(vars.email);
    });

    test("Changes password correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <LoginForm {...props} />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Password");
        fireEvent.change(input, { target: { value: vars.password } });
        expect(input).toHaveValue(vars.password);
    });

    test("Handles successful submit correctly", async () => {
        window.location.reload = jest.fn();
        const { getByRole, getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <LoginForm {...props} />
            </MockedProvider>,
        );
        await wait();

        const email = getByLabelText("Email");
        fireEvent.change(email, { target: { value: vars.email } });

        const password = getByLabelText("Password");
        fireEvent.change(password, { target: { value: vars.password } });

        const form = getByRole("form");
        fireEvent.submit(form);
        await wait();
        expect(props.onSubmitChange).toHaveBeenCalledTimes(3);
        expect(props.onSubmitChange).toHaveBeenNthCalledWith(2, true);
        expect(props.onSubmitChange).toHaveBeenNthCalledWith(3, false);
        expect(numTimesSubmitted).toEqual(1);
        expect(window.location.reload).toHaveBeenCalledTimes(1);
    });

    test("Handles successful submit correctly without onSubmitChange prop", async () => {
        window.location.reload = jest.fn();
        const { getByRole, getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <LoginForm {...props} onSubmitChange={undefined} />
            </MockedProvider>,
        );
        await wait();

        const email = getByLabelText("Email");
        fireEvent.change(email, { target: { value: vars.email } });

        const password = getByLabelText("Password");
        fireEvent.change(password, { target: { value: vars.password } });

        const form = getByRole("form");
        fireEvent.submit(form);
        await wait();
        expect(props.onSubmitChange).toHaveBeenCalledTimes(0);
        expect(numTimesSubmitted).toEqual(1);
    });

    test("Handles successful submit correctly from /login", async () => {
        window.history.pushState({}, 'Login', '/login');
        window.location.assign = jest.fn();

        const { getByRole, getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <LoginForm {...props} />
            </MockedProvider>,
        );
        await wait();

        const email = getByLabelText("Email");
        fireEvent.change(email, { target: { value: vars.email } });

        const password = getByLabelText("Password");
        fireEvent.change(password, { target: { value: vars.password } });

        expect(window.location.pathname).toEqual("/login");

        const form = getByRole("form");
        fireEvent.submit(form);
        await wait();
        expect(props.onSubmitChange).toHaveBeenCalledTimes(3);
        expect(props.onSubmitChange).toHaveBeenNthCalledWith(2, true);
        expect(props.onSubmitChange).toHaveBeenNthCalledWith(3, false);
        expect(numTimesSubmitted).toEqual(1);
        expect(window.location.assign).toHaveBeenCalledTimes(1);
    });


    test("Handles submit with returned error correctly", async () => {
        const mockReturnError = {
            request: { query: LOGIN_USER, variables: { ...vars } },
            result: () => {
                numTimesSubmitted++;
                return {
                    data: { loginUser: { token: "", error: "invalid login..." } },
                };
            },
        };

        const { getByRole, getByText, getByLabelText } = render(
            <MockedProvider mocks={[mockReturnError]} addTypename={false}>
                <LoginForm {...props} />
            </MockedProvider>,
        );
        await wait();

        const email = getByLabelText("Email");
        fireEvent.change(email, { target: { value: vars.email } });

        const password = getByLabelText("Password");
        fireEvent.change(password, { target: { value: vars.password } });

        const form = getByRole("form");
        fireEvent.submit(form);
        await wait();
        expect(props.onSubmitChange).toHaveBeenCalledTimes(3);
        expect(props.onSubmitChange).toHaveBeenNthCalledWith(2, true);
        expect(props.onSubmitChange).toHaveBeenNthCalledWith(3, false);
        expect(numTimesSubmitted).toEqual(1);
        expect(getByText("invalid login...")).toBeDefined();
    });

    test("Handles submit with no returned token or error correctly", async () => {
        const mockReturnError = {
            request: { query: LOGIN_USER, variables: { ...vars } },
            result: () => {
                numTimesSubmitted++;
                return {
                    data: { loginUser: { token: "", error: "" } },
                };
            },
        };

        const { getByRole, getByText, getByLabelText } = render(
            <MockedProvider mocks={[mockReturnError]} addTypename={false}>
                <LoginForm {...props} />
            </MockedProvider>,
        );
        await wait();

        const email = getByLabelText("Email");
        fireEvent.change(email, { target: { value: vars.email } });

        const password = getByLabelText("Password");
        fireEvent.change(password, { target: { value: vars.password } });

        const form = getByRole("form");
        fireEvent.submit(form);
        await wait();
        expect(props.onSubmitChange).toHaveBeenCalledTimes(3);
        expect(props.onSubmitChange).toHaveBeenNthCalledWith(2, true);
        expect(props.onSubmitChange).toHaveBeenNthCalledWith(3, false);
        expect(numTimesSubmitted).toEqual(1);
        expect(getByText("something went wrong...")).toBeDefined();
    });

    test("Handles submit with network error correctly", async () => {
        const mockError = {
            request: { query: LOGIN_USER, variables: { ...vars } },
            error: new Error("something went wrong..."),
        };
        const { getByRole, getByText, getByLabelText } = render(
            <MockedProvider mocks={[mockError]} addTypename={false}>
                <LoginForm {...props} />
            </MockedProvider>,
        );
        await wait();

        const email = getByLabelText("Email");
        fireEvent.change(email, { target: { value: vars.email } });

        const password = getByLabelText("Password");
        fireEvent.change(password, { target: { value: vars.password } });

        const form = getByRole("form");
        fireEvent.submit(form);
        await wait();
        expect(props.onSubmitChange).toHaveBeenCalledTimes(3);
        expect(props.onSubmitChange).toHaveBeenNthCalledWith(2, true);
        expect(props.onSubmitChange).toHaveBeenNthCalledWith(3, false);
        expect(getByText("Network error: something went wrong...")).toBeDefined();
    });
});
