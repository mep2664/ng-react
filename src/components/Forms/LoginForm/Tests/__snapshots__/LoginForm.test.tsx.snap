// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the LoginForm component Component matches snapshot 1`] = `
<div>
  <form
    id="FORM_ID"
  >
    <div>
      Login
    </div>
    
    <label
      class="sc-iAyFgw fQEosV"
      for="textInput__Email"
    >
      Email
    </label>
    <input
      class="sc-hSdWYo dqOLGd"
      id="textInput__Email"
      name="email"
      type="text"
    />
    <label
      class="sc-iAyFgw fQEosV"
      for="textInput__Password"
    >
      Password
    </label>
    <input
      class="sc-hSdWYo dqOLGd"
      id="textInput__Password"
      name="password"
      type="text"
    />
  </form>
</div>
`;
